mm2src/coins/eth/wallet_connect.rs

```diff
@@ -0,0 +1,255 @@
+/// https://docs.reown.com/advanced/multichain/rpc-reference/ethereum-rpc
+use crate::common::Future01CompatExt;
+use crate::Eip1559Ops;
+use crate::{BytesJson, TransactionErr};
+
+use common::log::info;
+use derive_more::Display;
+use enum_derives::EnumFromStringify;
+use ethcore_transaction::{Action, SignedTransaction};
+use ethereum_types::H256;
+use ethereum_types::{Address, Public, H160, H520, U256};
+use ethkey::{public_to_address, Message, Signature};
+use kdf_walletconnect::WalletConnectOps;
+use kdf_walletconnect::{chain::{WcChainId, WcRequestMethods},
+                        error::WalletConnectError,
+                        WalletConnectCtx};
+use mm2_err_handle::prelude::*;
+use secp256k1::PublicKey;
+use secp256k1::{recovery::{RecoverableSignature, RecoveryId},
+                Secp256k1};
+use std::str::FromStr;
+use web3::signing::hash_message;
+
+use super::EthCoin;
+
+// Wait for 60 seconds for the transaction to appear on the RPC node.
+const WAIT_RPC_TIMEOUT_SECS: u64 = 60;
+
+#[derive(Display, Debug, EnumFromStringify)]
+pub enum EthWalletConnectError {
+    UnsupportedChainId(WcChainId),
+    InvalidSignature(String),
+    AccoountMisMatch(String),
+    #[from_stringify("rlp::DecoderError", "hex::FromHexError")]
+    TxDecodingFailed(String),
+    #[from_stringify("ethkey::Error")]
+    InternalError(String),
+    InvalidTxData(String),
+    WalletConnectError(WalletConnectError),
+}
+
+impl From<WalletConnectError> for EthWalletConnectError {
+    fn from(value: WalletConnectError) -> Self { Self::WalletConnectError(value) }
+}
+
+pub struct WcEthTxParams<'a> {
+    pub(crate) gas: U256,
+    pub(crate) nonce: U256,
+    pub(crate) data: &'a [u8],
+    pub(crate) my_address: H160,
+    pub(crate) action: Action,
+    pub(crate) value: U256,
+    pub(crate) gas_price: Option<U256>,
+}
+
+impl<'a> WcEthTxParams<'a> {
+    fn prepare_wc_tx_format(&self) -> MmResult<serde_json::Value, EthWalletConnectError> {
+        fn u256_to_hex(value: U256) -> String { format!("0x{:x}", value) }
+
+        let mut tx_json = json!({
+            "nonce": u256_to_hex(self.nonce),
+            "from": format!("0x{}", hex::encode(self.my_address.as_bytes())),
```

Comment:
Doesn't the display implementation of H160 already does that?

Reply:
done